R/generate_correlated_data.R

```diff
@@ -586,3 +586,115 @@ genCorOrdCat <- function(dtName, idname = "id", adjVar = NULL, baseprobs,
     asFactor = FALSE
   )
 }
+
+#' Create a block correlation matrix
+#' @description  The function genBlockMat() generates correlation matrices that 
+#' can accommodate clustered observations over time where the within-cluster 
+#' between-individual correlation in the same time period can be different from the 
+#' within-cluster between-individual correlation across time periods.The matrix
+#' generated here can be used in function addCorGen().
+#' @param rho The correlation coefficient between 0 and 1.
+#' @param nInds The number of units (individuals) in each period.
+#' @param nPeriods The number periods that data are observed.
+#' @param corstr Correlation structure for the matrix. Options include "ind" 
+#' for an independence structure (default), cs" for a compound symmetry 
+#' structure and "ar1" for an autoregressive structure.
+#' @param iRho Optional second correlation coefficient for within-individual across-period
+#' correlation in case the measurements for individuals are repeated over time. 
+#' This will use the same correlation structure as the between-individual correlation
+#' specified by "corstr".
+#' @return A block correlation matrix of size (nInds \* nPeriods) x (nInds \* nPeriods) 
+#' @examples
+#' genBlockMat(rho = 0.6, nInds = 2, nPeriods = 3, corstr = "ind")
+#' genBlockMat(rho = 0.6, nInds = 2, nPeriods = 3, corstr = "cs")
+#' genBlockMat(rho = 0.6, nInds = 3, nPeriods = 3, corstr = "ar1")
+#' 
+#' genBlockMat(rho = 0.6, nInds = 3, nPeriods = 3, corstr = "cs", iRho = 0.8)
+#' genBlockMat(rho = 0.6, nInds = 3, nPeriods = 3, corstr = "ar1", iRho = 0.8)
+#' @export
+#' @concept correlated
+genBlockMat <- function(rho, nInds, nPeriods, corstr = "ind", 
+                        iRho = NULL) {
+  
+  ### Checking
+  
+  assertNotMissing(rho = missing(rho), nInds = missing(nInds),
+                   nPeriods = missing(nPeriods))
+  
+  assertInteger(nInds = nInds, nPeriods = nPeriods)
+  assertAtLeast(nPeriods = nPeriods, minVal = 2)
+  assertInRange(rho = rho, range = c(-1,1))
+  
+  if (!is.null(iRho)) {
+    assertInRange(iRho = iRho, range = c(-1,1))
+    if (length(rho) == 1) {
+      assertNotEqual(corstr = corstr, val = "ind", 
+        msg = " With iRho set, corstr should be either cs or ar1.")  
+    }
+  }
+  
+  if (length(rho) > 1) {
+    assertLength(rho = rho, length = nPeriods)
+    if (!is.null(iRho)) {
+      assertLength(iRho = iRho, length = (nPeriods - 1))
+    }
+  }
+  
+  assertOption(corstr = corstr, options = c("ind", "cs", "ar1"))
+  
+  ###
+  
+  assignDiag <- function(block, value) {
+    diag(block) <- value
+    return(block)
+  }
```

Comment:
Not sure if this was your intention but I like how this function turns the "impure" (because it modifies the matrix in-place) `diag` function into a pure function :+1:

Reply:
I definitely had not been aware of the formal distinction between "pure" vs "impure" functions, but I have been wary of the kinds of functions that are impure, because it does make it feel like things are a bit out of control. The package "data.table" functions seem to have long lasting effects on data sets when used in a function. I don't quite see how what I've done here makes the overall genBlockMat pure, since I could have put assignDiag outside the function and I would not have been modifying anything outside the function addCorGen. I did it, because it seems to be a pretty specific case where I needed to use lapply with diag, and this was the only way I could get it to work properly. But, I'll take credit for being "pure". :).